lib/src/Pos/Diffusion/Full.hs

```diff
@@ -0,0 +1,445 @@
+import           Pos.Core.Coin (coinPortionToDouble)
+import           Pos.Core.Configuration (protocolMagic)
+import           Pos.Core.Types (HeaderHash, BlockVersionData (..), StakeholderId)
+import           Pos.Core.Txp (TxAux)
+import           Pos.Core.Update (UpId, UpdateVote, UpdateProposal)
+import           Pos.Crypto.Configuration (ProtocolMagic (..))
+import           Pos.DHT.Real (KademliaDHTInstance, KademliaParams (..), startDHTInstance, stopDHTInstance)
+import qualified Pos.Diffusion.Full.Block      as Diffusion.Block
+import qualified Pos.Diffusion.Full.Delegation as Diffusion.Delegation
+import qualified Pos.Diffusion.Full.Ssc        as Diffusion.Ssc
+import qualified Pos.Diffusion.Full.Txp        as Diffusion.Txp
+import qualified Pos.Diffusion.Full.Update     as Diffusion.Update
+import           Pos.Diffusion.Full.Types  (DiffusionWorkMode)
+import           Pos.Diffusion.Types (Diffusion (..), DiffusionLayer (..), GetBlocksError (..))
+import           Pos.Logic.Types (Logic (..))
+import           Pos.Network.CLI (NetworkConfigOpts (..), intNetworkConfigOpts)
+import           Pos.Network.Types (NetworkConfig (..), Topology (..), Bucket, initQueue,
+                                    topologySubscribers, SubscriptionWorker (..),
+                                    topologySubscriptionWorker)
+import           Pos.Ssc.Message (MCOpening, MCShares, MCCommitment, MCVssCertificate)
+import           Pos.Subscription.Common (subscriptionListeners)
+import           Pos.Update.Configuration (lastKnownBlockVersion)
+import           Pos.Util.OutboundQueue (EnqueuedConversation (..))
+
+-- | The full diffusion layer.
+--
+-- NB: we could do the whole logic/diffusion layer interface using typeclasses
+-- against a monad, but we'd end up with a bunch of reader constraints over
+-- the values that would otherwise appear in the Logic and Diffusion records.
+-- That's to say, we'd have to do the same work anyway, but then even more
+-- work to juggle the instances.
+diffusionLayerFull
+    :: forall d x .
+       ( DiffusionWorkMode d
+       , MonadFix d
+       )
+    => NetworkConfigOpts
+    -> Maybe Monitoring.Store
+    -> ((Logic d -> Production (DiffusionLayer d)) -> Production x)
+    -> Production x
+diffusionLayerFull networkConfigOpts mEkgStore expectLogic =
+    bracket acquire release $ \_ -> expectLogic $ \logic -> do
+
+        -- Read in network topology / configuration / policies.
+        networkConfig <- intNetworkConfigOpts networkConfigOpts
+
+        -- Make the outbound queue using network policies.
+        oq :: OQ.OutboundQ (EnqueuedConversation d) NodeId Bucket <-
+            initQueue networkConfig mEkgStore 
+
+        let -- VerInfo is a diffusion-layer-specific thing. It's only used for
+            -- negotiating with peers.
+            ourVerInfo :: VerInfo
+            ourVerInfo = VerInfo (getProtocolMagic protocolMagic) lastKnownBlockVersion ins (outs <> workerOuts)
```

Comment:
Block version can and will change while diffusion layer is running. How do we deal with that fact?

Reply:
Oh interesting point. I'm not sure if we have ever dealt with that! `VerInfo` is and always has been constant for the life time of the process.